src/main/scala/no/ndla/draftapi/integration/TaxonomyApiClient.scala

```diff
@@ -0,0 +1,157 @@
+/*
+ * Part of NDLA draft-api.
+ * Copyright (C) 2018 NDLA
+ *
+ * See LICENSE
+ */
+
+package no.ndla.draftapi.integration
+import java.util.concurrent.Executors
+
+import cats.Traverse
+import com.typesafe.scalalogging.LazyLogging
+import no.ndla.network.{AuthUser, NdlaClient}
+import no.ndla.network.model.HttpRequestException
+import no.ndla.draftapi.DraftApiProperties.Domain
+import no.ndla.draftapi.model.domain.{Article, ArticleTitle, Language}
+import org.json4s.{DefaultFormats, Formats}
+import scalaj.http.Http
+import org.json4s.jackson.Serialization.write
+
+import scala.util.{Failure, Success, Try}
+import cats.implicits._
+
+trait TaxonomyApiClient {
+  this: NdlaClient =>
+  val taxonomyApiClient: TaxonomyApiClient
+
+  class TaxonomyApiClient extends LazyLogging {
+    private val TaxonomyApiEndpoint = s"$Domain/taxonomy/v1"
+    private val taxonomyTimeout = 20 * 1000 // 20 Seconds
+    implicit val formats: DefaultFormats.type = DefaultFormats
+
+    def queryResource(articleId: Long): Try[List[Resource]] = {
+      get[List[Resource]](s"$TaxonomyApiEndpoint/queries/resources", "contentURI" -> s"urn:article:$articleId")
+    }
+
+    def queryTopic(articleId: Long): Try[List[Topic]] = {
+      get[List[Topic]](s"$TaxonomyApiEndpoint/queries/topics", "contentURI" -> s"urn:article:$articleId")
+    }
+
+    def updateTaxonomyIfExists(articleId: Long, article: Article): Try[Long] = {
```

Comment:
Synes denne ble litt i største laget. Tror den kan bli vesentlig enklere hvis du splitter den opp litt.
Foreslå å lage en enkelt funksjon som oppdaterer enten en ressurs eller en topic. F.eks `private def updateTaxonomyTitle(title: String, updateFunction: (String, Any) => Try[_]) ...`.
Denne funksjonen kunne da klart seg med å gjøre noe ala dette
```scala
for {
    a <- resources.map(r => updateTaxonomyTitle(r.copy, updateResource, article)
    b <- topics.map(t => updateTaxonomyTitle(t.copy, updateTopic, article)
}
```

Reply:
Jeg skjønner ikke hvordan `updateTaxonomyIfExists` funksjonen blir mindre om jeg gjør dette? 
Men har pushet en commit som fikset så jeg sender inn funksjonen for å oppdatere.

Reply:
Tenkte mere på noe som dette. Ikke testa om det funker, men _tror_ kanskje idéen er sound :)
```scala
    def updateTaxonomyIfExists(articleId: Long, article: Article): Try[Long] = {
      for {
        resources <- queryResource(articleId)
        _ <- updateTaxonomyIfExists[Resource](resources, article.title, updateResourceTitleAndTranslations)
        topics <- queryTopic(articleId)
        _ <- updateTaxonomyIfExists[Topic](topics, article.title, updateTopicTitleAndTranslations)
      } yield articleId
    }

    private def updateTaxonomyIfExists[T](resource: Seq[T], titles: Seq[ArticleTitle], updateFunc: (T, ArticleTitle, Seq[ArticleTitle]) => Try[T]): Try[List[T]] = {
      Language.findByLanguageOrBestEffort(titles, Language.DefaultLanguage) match {
        case Some(title) =>
          val a = resource.map(updateFunc(_, title, titles))
           a.collect { case Failure(ex) => ex }
              .foreach(ex => logger.warn(s"Taxonomy update failed with: ${ex.getMessage}"))

          Traverse[List].sequence(a.toList)
        case None => Failure(new RuntimeException("This is a bug, no name was found for published article."))
      }
    }

    private def updateResourceTitleAndTranslations(res: Resource, defaultTitle: ArticleTitle, titles: Seq[ArticleTitle]): Try[Resource] =
      updateTitleAndTranslations(res, defaultTitle, titles, updateResource, updateResourceTranslations(res.id, _))

    private def updateTopicTitleAndTranslations(top: Topic, defaultTitle: ArticleTitle, titles: Seq[ArticleTitle]): Try[Topic] =
      updateTitleAndTranslations(top, defaultTitle, titles, updateTopic, updateTopicTranslations(top.id, _))

    private def updateTitleAndTranslations[T <: Taxonomy[T]](res: T, defaultTitle: ArticleTitle, titles: Seq[ArticleTitle], update: T => Try[T], updateTranslations: Seq[ArticleTitle] => Try[List[Translation]]) = {
      val resourceResult = update(res.withName(defaultTitle.title))
      val translationResult = updateTranslations(res.id, titles)

      (resourceResult, translationResult) match {
        case (Success(s1), Success(_)) => Success(s1)
        case (Failure(ex), _)          => Failure(ex)
        case (_, Failure(ex))          => Failure(ex)
      }
    }
```

Reply:
Det virka smart :+1: